File: Listas_Duplamente_Encadeadas.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int value;
    struct Node* prev;
    struct Node* next;
} Node;

typedef struct {
    Node* head;
    Node* tail;
    int size;
} DList;

DList* create_list() {
    DList* list = (DList*)malloc(sizeof(DList));
    list->head = list->tail = NULL;
    list->size = 0;
    return list;
}

void push_front(DList* list, int value) {
    Node* n = (Node*)malloc(sizeof(Node));
    n->value = value; n->prev = NULL; n->next = list->head;
    if (list->head) list->head->prev = n; else list->tail = n;
    list->head = n; list->size++;
}

void push_back(DList* list, int value) {
    Node* n = (Node*)malloc(sizeof(Node));
    n->value = value; n->next = NULL; n->prev = list->tail;
    if (list->tail) list->tail->next = n; else list->head = n;
    list->tail = n; list->size++;
}

Node* find(DList* list, int value) {
    Node* cur = list->head;
    while (cur) {
        if (cur->value == value) return cur;
        cur = cur->next;
    }
    return NULL;
}

void remove_node(DList* list, Node* n) {
    if (!n) return;
    if (n->prev) n->prev->next = n->next; else list->head = n->next;
    if (n->next) n->next->prev = n->prev; else list->tail = n->prev;
    free(n); list->size--;
}

void print_list(DList* list) {
    Node* cur = list->head;
    while (cur) {
        printf("%d ", cur->value);
        cur = cur->next;
    }
    printf("\n");
}

int main() {
    DList* lista = create_list();
    push_back(lista, 10);
    push_front(lista, 5);
    push_back(lista, 20);
    print_list(lista); 

    return 0;
}

File: Listas_Duplamente_Encadeadas.py
class Node:
    def __init__(self, value):
        self.value = value
        self.prev = None
        self.next = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def is_empty(self):
        return self.size == 0

    def push_front(self, value):
        new_node = Node(value)
        if self.is_empty():
            self.head = self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node
        self.size += 1

    def push_back(self, value):
        new_node = Node(value)
        if self.is_empty():
            self.head = self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node
        self.size += 1

    def pop_front(self):
        if self.is_empty():
            return None
        value = self.head.value
        self.head = self.head.next
        if self.head:
            self.head.prev = None
        else:
            self.tail = None
        self.size -= 1
        return value

    def pop_back(self):
        if self.is_empty():
            return None
        value = self.tail.value
        self.tail = self.tail.prev
        if self.tail:
            self.tail.next = None
        else:
            self.head = None
        self.size -= 1
        return value

    def find(self, value):
        cur = self.head
        while cur:
            if cur.value == value:
                return cur
            cur = cur.next
        return None

    def remove(self, node):
        if node is None:
            return False
        if node.prev:
            node.prev.next = node.next
        else:
            self.head = node.next
        if node.next:
            node.next.prev = node.prev
        else:
            self.tail = node.prev
        self.size -= 1
        return True

    def print_list(self):
        cur = self.head
        elementos = []
        while cur:
            elementos.append(str(cur.value))
            cur = cur.next
        print(" <-> ".join(elementos))

# Exemplo de uso:
if __name__ == "__main__":
    dll = DoublyLinkedList()
    dll.push_back(10)
    dll.push_front(5)
    dll.push_back(20)
    dll.print_list() 